### PR TITLE
Stop validators

### PR DESCRIPTION
**Pull request by pieter-edelman-nictiz**
#### Brief overview of PR changes/additions
Make the Tasks that perform the validation cancellable, and make the Java validator Process killable.

#### Motivation for adding to Hammer
See issue #62

### CODE REVIEW

**Comment by vadi2** [comment]
Ooh... first time I've seen that complain! Want to fix that?

**Comment by pieter-edelman-nictiz** [comment]
Just made some changes ... let's see if it works.

File: Program.cs [hunk 1]
```diff
@@ -177,6 +178,10 @@ public ValidationResult DotnetResult
       set => this.SetProperty(ref _dotnetResult, value);
     }
 
+    private CancellationTokenSource validatorCancellationSource = null;
```
Comment (vadi2) [inline Program.cs]:
```suggestion
    private CancellationTokenSource _validatorCancellationSource;
```

It seems the C# convention is to start private fields with an underscore.

Don't have to set it to `null`, C# runtime does that by default (different from C++)

File: Program.cs [hunk 2]
```diff
@@ -177,6 +178,10 @@ public ValidationResult DotnetResult
       set => this.SetProperty(ref _dotnetResult, value);
     }
 
+    private CancellationTokenSource validatorCancellationSource = null;
+    
+    private List<Process> validatorProcesses = new List<Process>();
```
Comment (vadi2) [inline Program.cs]:
```suggestion
    private List<Process> _validatorProcesses = new List<Process>();
```

Same

File: Program.cs [hunk 3]
```diff
@@ -582,44 +595,71 @@ public OperationOutcome ValidateWithJava()
...
+      if (validatorCancellationSource != null) {
+        validatorCancellationSource.Cancel();
+        validatorCancellationSource.Dispose();
+      }
+      validatorCancellationSource = null;
+      
+      // We can actively kill the Java validator as this is an external
+      // process. The .NET validator needs to run its course until completion,
+      // we'll just ignore the results.
+      foreach (Process process in validatorProcesses) {
```
Comment (vadi2) [inline Program.cs]:
```suggestion
      foreach (var process in validatorProcesses) {
```

Redundant to qualify it here

**Comment by vadi2** [comment]
Can I help with the conflict resolution?

**Comment by pieter-edelman-nictiz** [comment]
I don't know, I keep running into the same merge conflicts. Hope to have it fixed soon ...

**Comment by vadi2** [comment]
Remember you need to delete the <<<< lines and such:

```js
<<<<<<< HEAD
<<<<<<< HEAD
<<<<<<< HEAD
                        appmodel.cancelValidation()
                    }
=======
>>>>>>> Implement validator cancelation
=======
                        appmodel.cancelValidation()
                    }
>>>>>>> Implement validator cancelation
=======
                        appmodel.cancelValidation()
                    }
>>>>>>> 15a26ad6d0aafb3a73e542c2c4f837caa83c7201
````

**Comment by pieter-edelman-nictiz** [comment]
Yeah, I missed them. Should be fixed now

**Comment by pieter-edelman-nictiz** [comment]
No, not fixed just yet ...

**Comment by pieter-edelman-nictiz** [comment]
Ok, I really think it should be all ok now

**Comment by vadi2** [comment]
If the Java validator is validating and I hit re-validate, the busy spinner doesn't show again - it says no errors - but it's really validating in the background, and comes back again... do you have this?

**Comment by vadi2** [comment]
Ah I think you're just missing a `appmodel.cancelValidation()` call for the new instance validator, that's what.

**Comment by pieter-edelman-nictiz** [comment]
No, I think it's a concurreny issue. I'll look into it later today

**Comment by vadi2** [comment]
OK, but check my theory first, easier to test. See how we call `appmodel.startValidation()` if `(addResourcesPage.state === "VALIDATION_RESULTS" && resultsPageEditor.state === "VISIBLE")` - we never call cancel validation firstly.

**Comment by pieter-edelman-nictiz** [comment]
True, but CancelValidation() is called from StartValidation()

**Comment by vadi2** [comment]
Aha. Yes.

**Comment by pieter-edelman-nictiz** [comment]
Ok, please try again

**Comment by pieter-edelman-nictiz** [comment]
Yeah, glad it's finally fixed

**Pull request merged**